### DISCUSSION TITLE
Propeller physics - simulation of thrust/lift?

### DISCUSSION THREAD

**Discussion by lkoelman** [microsoft/AirSim#3356] [Support Q&A] [answered]
I am trying to find out how the propellor rotation is linked to the drone's movement/kinematics inside the Unreal engine.

It's not clear for me from the code (MultiRotor cpp classes) and BP_FlyingPawn blueprint how the propellor thrust vectors are passed on to the unreal engine. I see some code related to the calculation and storage of thrust/kinematic parameters in the following classes:

- MultiRotorPawnSimApi
- MultiRotorParams.hpp
- RotorActuator.hpp

However, I can't find the 'communication stage' between the internal physics model and the unreal engine. Or could it be that thrust is calculated through Unreal physics simulation based on the movement of the realistically shaped propellors?

**Comment by rajat2004** [comment]
AFAIK, the Unreal Engine copter model isn't associated with the physics, the model is used for the rendering of the copter, props, etc but the physics is all handled inside FastPhysicsEngine only.
In Car vehicle, the physics is handled inside UE using PhysX

**Comment by EdwinMeriaux** [comment]
@rajat2004 
So when designing our own custom drone is there anything to modify in the FastPhysicsEngine?

**Reply by jonyMarino** [reply]
If you need to simulate the behavior of a different dynamic, yes. Maybe modifying the parameter values can be sufficient for you. But it may fall short if you need precision for a very different drone and would need to modify the actual code.

**Comment by jonyMarino** [comment, marked as answer]
Hi @lkoelman!
As Rajat pointed out, there is no passage of physical values to UE other than the estimated position.
The physics are estimated on the World.hpp::WorldUpdatorAsync::update() thread and the estimated position is obtained through 
phys_body.getPose() on the SimModeWorldBase::Tick() thread, that takes care of the rendering.
